### PR TITLE
MODORDSTOR-45 - Build API for PO line numbers

### PR DESCRIPTION
**Pull request by khandramai**
## Purpose
Creation of API end point to generate PO line number.

## Approach
- New API implemented: GET /orders-storage/po-line-number?purchaseOrderId=UUID_of_PO.
- For existing APIs POST /orders-storage/purchase-orders and PUT /orders-storage/puchase-orders new operations for creating and deleting sequence at the database level added in such a way that within a single POST/ or PUT request operations **save PO/create Sequence** or **update PO/delete Sequence** are carried out respectively.

#### TODOS and Open Questions
It is necessary to consider/provide support of transactionality for operations:
- save PO/create Sequence (POST /orders-storage/purchase-orders);
- update PO/delete Sequence (PUT /orders-storage/puchase-orders).

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by khandramai** [comment]
JIC, SonarQube code issues will be fixed after final decision about transactionality (TODOs will be removed).

File: src/main/java/org/folio/rest/utils/HelperUtils.java [hunk 1]
```diff
@@ -21,5 +24,31 @@ public static void respond(Handler<AsyncResult<Response>> handler, Response resp
     handler.handle(result);
   }
 
-
+  public enum SequenceQuery {
+
+    CREATE_SEQUENCE {
+      @Override
+      public String getQuery(String purchaseOrderId) {
+        return "CREATE SEQUENCE " + constructSequenceName(purchaseOrderId) + " MINVALUE 1 MAXVALUE 999";
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/utils/HelperUtils.java]: See the comments in https://wiki.folio.org/display/FOLIJET/Postgres+SEQUENCEs+for+generating+po_line_number+suffixes

I think we probably want to be safe and add an IF NOT EXISTS clause.

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 2]
```diff
@@ -283,4 +283,43 @@ public void putOrdersStoragePurchaseOrdersById(String id, String lang, org.folio
...
+  private void processDropSequence(String lang, PurchaseOrder entity, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext,
+                                   String tenantId, AsyncResult<UpdateResult> reply) {
+    PurchaseOrder.WorkflowStatus status = entity.getWorkflowStatus();
+    if(status == PurchaseOrder.WorkflowStatus.OPEN || status == PurchaseOrder.WorkflowStatus.CLOSED) {
+      PostgresClient.getInstance(vertxContext.owner(), tenantId).select(DROP_SEQUENCE.getQuery(entity.getId()), sequenceDeleteReply -> {
+        if (sequenceDeleteReply.succeeded()) {
+          asyncResultHandler.handle(Future.succeededFuture(PutOrdersStoragePurchaseOrdersByIdResponse
+            .respond204()));
+        } else {
+          // TODO Transactionality (???) --> Roll-back PO update?
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: I'd say leave the TODO until we can figure out a better way to handle these failures

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 3]
```diff
@@ -283,4 +283,43 @@ public void putOrdersStoragePurchaseOrdersById(String id, String lang, org.folio
...
+                                     String tenantId, String persistenceId, OutStream stream) {
+    PostgresClient.getInstance(vertxContext.owner(), tenantId).execute(CREATE_SEQUENCE.getQuery(entity.getId()),
+      createSequenceReply -> {
+        if(createSequenceReply.succeeded()) {
+          Response response = PostOrdersStoragePurchaseOrdersResponse
+            .respond201WithApplicationJson(stream, PostOrdersStoragePurchaseOrdersResponse.headersFor201()
+              .withLocation(PURCHASE_ORDER_LOCATION_PREFIX + persistenceId));
+          respond(asyncResultHandler, response);
+        } else {
+          // TODO Transactionality (???) --> Roll-back PO creation?
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: I'd say leave the TODO until we can figure out a better way to handle these failures

File: src/main/java/org/folio/rest/utils/HelperUtils.java (hunk 1 above)
Comment (khandramai) [reply src/main/java/org/folio/rest/utils/HelperUtils.java]: @craigmcnally,

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java (hunk 2 above)
Comment (piotr-kalashuk) [reply src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: Please have a look at tips provided by Folijet team on the possible approach to work with transaction: [Chained futures](https://wiki.folio.org/display/FOLIJET/Chained+futures+and+rx-java+overview#Chainedfuturesandrx-javaoverview-Chainedfutures)

Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: Ah, thanks for pointing this out to me @piotr-kalashuk

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 4]
```diff
@@ -111,51 +111,53 @@ public void getOrdersStoragePurchaseOrders(String query, int offset, int limit,
   @Override
   @Validate
   public void postOrdersStoragePurchaseOrders(String lang, org.folio.rest.jaxrs.model.PurchaseOrder entity,
-      Map<String, String> okapiHeaders,
-      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+                                              Map<String, String> okapiHeaders,
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: this massive indentation doesn't look right to me

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 5]
```diff
@@ -246,80 +248,64 @@ public void deleteOrdersStoragePurchaseOrdersById(String id, String lang, Map<St
   @Override
   @Validate
   public void putOrdersStoragePurchaseOrdersById(String id, String lang, org.folio.rest.jaxrs.model.PurchaseOrder entity,
-      Map<String, String> okapiHeaders,
-      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+                                                 Map<String, String> okapiHeaders,
```
Comment (craigmcnally) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: this massive indentation doesn't look right to me

**Comment by craigmcnally** [comment]
test coverage on new code is below 80%, but that might change after resolving the merge conflicts.

**Comment by craigmcnally** [comment]
The test coverage is pretty close (78% or so).  I think ideally we'd bring that up a little bit by exercising the transaction rollback code, but I'm fine with creating a follow-on story to address that, and to lower the code complexity of the methods that use the transaction stuff...  See https://sonarcloud.io/project/issues?id=org.folio%3Amod-orders-storage&issues=AWjM897FeedjntrDprkt&open=AWjM897FeedjntrDprkt&pullRequest=79

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 6]
```diff
@@ -39,16 +50,42 @@ public void getOrdersStoragePurchaseOrders(String query, int offset, int limit,
...
-    PgUtil.post(PURCHASE_ORDER_TABLE, entity, okapiHeaders, vertxContext, PostOrdersStoragePurchaseOrdersResponse.class, asyncResultHandler);
+  public void postOrdersStoragePurchaseOrders(String lang, org.folio.rest.jaxrs.model.PurchaseOrder entity, Map<String, String> okapiHeaders,
+    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    try {
+      vertxContext.runOnContext(v -> {
+        Tx<PurchaseOrder> tx = new Tx<>(entity);
+        Future.succeededFuture(tx)
+          .compose(this::startTx)
+          .compose(this::createSequence)
+          .compose(t -> createPurchaseOrder(tx, entity, okapiHeaders, vertxContext))
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: The `Tx` holds already `entity` so makes sense to remove second param from `createPurchaseOrder`

File: src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java [hunk 7]
```diff
@@ -59,11 +96,133 @@ public void deleteOrdersStoragePurchaseOrdersById(String id, String lang, Map<St
...
-    PgUtil.put(PURCHASE_ORDER_TABLE, entity, id, okapiHeaders, vertxContext, PutOrdersStoragePurchaseOrdersByIdResponse.class, asyncResultHandler);
+  public void putOrdersStoragePurchaseOrdersById(String id, String lang, org.folio.rest.jaxrs.model.PurchaseOrder entity, Map<String, String> okapiHeaders,
+    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    try {
+      vertxContext.runOnContext(v -> {
+        Tx<PurchaseOrder> tx = new Tx<>(entity);
+        Future.succeededFuture(tx)
+          .compose(this::startTx)
+          .compose(this::deleteSequence)
+          .compose(t -> updatePurchaseOrder(tx, id, entity, okapiHeaders, vertxContext))
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/PurchaseOrdersAPI.java]: The `Tx` holds already `entity` so makes sense to remove third param from `updatePurchaseOrder`

**Pull request merged**